File: backend/mqtt_client.py
"""ClockMQTT Broker Client — Paho MQTT wrapper.

Handles communication between the Web Console and Aedes MQTT broker.
Both run on the same server, so broker is at localhost:2082.
"""

from __future__ import annotations

import json
import logging
import threading
from typing import Callable, Optional

import paho.mqtt.client as mqtt

logger = logging.getLogger("mqtt_client")

# ============================================================================
# Config
# ============================================================================

MQTT_BROKER_HOST = "127.0.0.1"
MQTT_BROKER_PORT = 2082
MQTT_USERNAME = "admin"
MQTT_PASSWORD = "admin123"  # CHANGE IN PRODUCTION

# ============================================================================
# Client
# ============================================================================


class ClockMQTTClient:
    """Singleton MQTT client for the Web Console backend."""

    _instance: Optional["ClockMQTTClient"] = None

    def __new__(cls) -> "ClockMQTTClient":
        if cls._instance is None:
            cls._instance = super().__new__(cls)
        return cls._instance

    def __init__(self) -> None:
        if hasattr(self, "_initialized"):
            return
        self._initialized = True

        self._client = mqtt.Client(mqtt.CallbackAPIVersion.VERSION2)
        self._client.username_pw_set(MQTT_USERNAME, MQTT_PASSWORD)
        self._client.on_connect = self._on_connect
        self._client.on_message = self._on_message

        self._handlers: dict[str, list[Callable]] = {}
        self._connected = False
        self._lock = threading.Lock()

    # ------------------------------------------------------------------
    # Connection
    # ------------------------------------------------------------------

    def connect(self) -> None:
        """Connect to the MQTT broker."""
        if self._connected:
            return
        logger.info(f"Connecting to MQTT broker at {MQTT_BROKER_HOST}:{MQTT_BROKER_PORT}")
        self._client.connect_async(MQTT_BROKER_HOST, MQTT_BROKER_PORT)
        self._client.loop_start()

    def disconnect(self) -> None:
        """Disconnect from the MQTT broker."""
        self._client.loop_stop()
        self._client.disconnect()
        self._connected = False

    @property
    def is_connected(self) -> bool:
        return self._connected

    # ------------------------------------------------------------------
    # Callbacks
    # ------------------------------------------------------------------

    def _on_connect(self, client, userdata, flags, reason_code, properties):
        if reason_code == 0:
            self._connected = True
            logger.info("MQTT connected OK")
            # Subscribe to all device status topics
            self._client.subscribe("inkpad/+/status")
        else:
            logger.error(f"MQTT connect failed: rc={reason_code}")

    def _on_message(self, client, userdata, msg):
        try:
            payload = msg.payload.decode("utf-8")
            logger.debug(f"MQTT RX: {msg.topic} → {payload[:100]}")
        except Exception:
            return

        # Dispatch to registered handlers
        with self._lock:
            for pattern, handlers in self._handlers.items():
                if mqtt.topic_matches_sub(pattern, msg.topic):
                    for handler in handlers:
                        try:
                            handler(msg.topic, payload)
                        except Exception as e:
                            logger.error(f"Handler error for {msg.topic}: {e}")

    def on_topic(self, pattern: str, handler: Callable) -> None:
        """Register a handler for a topic pattern."""
        with self._lock:
            self._handlers.setdefault(pattern, []).append(handler)

    # ------------------------------------------------------------------
    # Publish helpers
    # ------------------------------------------------------------------

    def publish(self, topic: str, payload: dict | str, qos: int = 1, retain: bool = False) -> None:
        """Publish a message to a topic."""
        if isinstance(payload, dict):
            payload = json.dumps(payload, ensure_ascii=False)
        info = self._client.publish(topic, payload, qos=qos, retain=retain)
        logger.debug(f"MQTT TX: {topic} → {str(payload)[:100]}")

    def publish_time_sync(self, device_id: str, timestamp: int) -> None:
        self.publish(
            f"inkpad/{device_id}/time/sync",
            {"timestamp": timestamp, "timezone": "Asia/Shanghai"},
            qos=0,
        )

    def publish_word(self, device_id: str, word_data: dict) -> None:
        self.publish(
            f"inkpad/{device_id}/word/daily",
            word_data,
            qos=1,
            retain=True,
        )

    def publish_schedule(self, device_id: str, schedules: list[dict]) -> None:
        self.publish(
            f"inkpad/{device_id}/schedule/update",
            {"version": 1, "schedules": schedules},
            qos=1,
            retain=True,
        )

    def publish_display_text(
        self, device_id: str, lines: list[dict], duration_sec: int = 30
    ) -> None:
        self.publish(
            f"inkpad/{device_id}/display/text",
            {"lines": lines, "duration_sec": duration_sec},
            qos=1,
        )

    def publish_config(self, device_id: str, config: dict) -> None:
        self.publish(
            f"inkpad/{device_id}/config/set",
            config,
            qos=1,
        )


# ============================================================================
# Singleton accessor
# ============================================================================

mqtt_broker = ClockMQTTClient()

File: backend/mqttClient.js
/**
 * ClockMQTT — MQTT Client
 * ========================
 * Singleton mqtt.js wrapper. Replaces the Python Paho MQTT client.
 *
 * Unlike Paho's thread-based model, mqtt.js integrates with the Node.js
 * event loop — subscriptions in the 'connect' callback are reliably active
 * before any messages arrive, eliminating the Paho race condition.
 */

const mqtt = require('mqtt');

// ============================================================================
// Config
// ============================================================================

const MQTT_BROKER_URL  = 'mqtt://127.0.0.1:2082';
const MQTT_USERNAME    = 'admin';
const MQTT_PASSWORD    = 'admin123'; // CHANGE IN PRODUCTION
const MQTT_CLIENT_ID   = 'clockmqtt_web_backend';

// ============================================================================
// Topic Pattern Matching
// ============================================================================

/**
 * Check if an MQTT topic matches a subscription pattern.
 * Supports single-level (+) and multi-level (#) wildcards.
 * # must appear only at the end of the pattern.
 */
function topicMatchesPattern(pattern, topic) {
  const patParts = pattern.split('/');
  const topParts = topic.split('/');

  for (let i = 0; i < patParts.length; i++) {
    if (patParts[i] === '#') {
      // Multi-level wildcard — matches everything remaining (including nothing)
      return i === patParts.length - 1; // valid only at end
    }
    if (patParts[i] === '+') {
      // Single-level wildcard — matches exactly one level
      if (i >= topParts.length) return false;
      continue;
    }
    if (i >= topParts.length || patParts[i] !== topParts[i]) {
      return false;
    }
  }
  return patParts.length === topParts.length;
}

// ============================================================================
// Client
// ============================================================================

class ClockMQTTClient {
  constructor() {
    this._client = null;
    this._connected = false;
    this._handlers = new Map(); // pattern -> [handlerFn]
  }

  // ------------------------------------------------------------------
  // Connection
  // ------------------------------------------------------------------

  connect() {
    if (this._client) return;

    console.log(`[MQTT] Connecting to ${MQTT_BROKER_URL} as "${MQTT_USERNAME}"...`);
    this._client = mqtt.connect(MQTT_BROKER_URL, {
      username: MQTT_USERNAME,
      password: MQTT_PASSWORD,
      clientId: MQTT_CLIENT_ID,
      clean: true,
      reconnectPeriod: 5000, // auto-reconnect every 5s
    });

    this._client.on('connect', () => {
      this._connected = true;
      console.log('[MQTT] Connected OK');

      // Subscribe to device status and presence topics
      this._client.subscribe(['inkpad/+/status', 'inkpad/+/presence'], { qos: 1 }, (err) => {
        if (err) {
          console.error('[MQTT] Subscribe error:', err.message);
        } else {
          console.log('[MQTT] Subscribed to inkpad/+/status and inkpad/+/presence');
        }
      });
    });

    this._client.on('reconnect', () => {
      console.log('[MQTT] Reconnecting...');
    });

    this._client.on('close', () => {
      this._connected = false;
      console.log('[MQTT] Connection closed');
    });

    this._client.on('error', (err) => {
      console.error('[MQTT] Error:', err.message);
    });

    this._client.on('message', (topic, payload) => {
      let text;
      try {
        text = payload.toString();
      } catch {
        console.warn(`[MQTT] RX: cannot decode payload on ${topic}`);
        return;
      }

      console.log(`[MQTT] RX: ${topic} → ${text.slice(0, 150)}`);

      // Dispatch to registered handlers
      let matched = false;
      for (const [pattern, handlers] of this._handlers) {
        if (topicMatchesPattern(pattern, topic)) {
          matched = true;
          for (const handler of handlers) {
            try {
              handler(topic, text);
            } catch (e) {
              console.error(`[MQTT] Handler error for ${topic}:`, e.message);
            }
          }
        }
      }
      if (!matched) {
        console.warn(`[MQTT] RX: no handler for ${topic} (patterns: ${[...this._handlers.keys()].join(', ')})`);
      }
    });
  }

  disconnect() {
    if (this._client) {
      this._client.end(true);
      this._client = null;
    }
    this._connected = false;
  }

  get isConnected() {
    return this._connected;
  }

  // ------------------------------------------------------------------
  // Handler registration
  // ------------------------------------------------------------------

  onTopic(pattern, handler) {
    if (!this._handlers.has(pattern)) {
      this._handlers.set(pattern, []);
    }
    this._handlers.get(pattern).push(handler);
    console.log(`[MQTT] Registered handler for "${pattern}"`);
  }

  // ------------------------------------------------------------------
  // Publish helpers
  // ------------------------------------------------------------------

  publish(topic, payload, qos = 1, retain = false) {
    if (!this._client || !this._connected) {
      console.warn(`[MQTT] TX skipped (not connected): ${topic}`);
      return;
    }
    const data = typeof payload === 'object' ? JSON.stringify(payload) : payload;
    this._client.publish(topic, data, { qos, retain }, (err) => {
      if (err) console.error(`[MQTT] TX error ${topic}:`, err.message);
    });
    console.log(`[MQTT] TX: ${topic} → ${data.slice(0, 100)}`);
  }

  publishTimeSync(deviceId, timestamp) {
    const now = new Date();
    this.publish(
      `inkpad/${deviceId}/time/sync`,
      {
        timestamp,
        timezone: 'Asia/Shanghai',
        datetime: now.toISOString(),
      },
      0, // qos=0
      false,
    );
  }

  publishWord(deviceId, wordData) {
    this.publish(
      `inkpad/${deviceId}/word/daily`,
      wordData,
      1,    // qos=1
      true, // retain
    );
  }

  publishSchedule(deviceId, schedules) {
    this.publish(
      `inkpad/${deviceId}/schedule/update`,
      { version: 1, schedules: Array.isArray(schedules) ? schedules : [schedules] },
      1,    // qos=1
      true, // retain
    );
  }

  publishDisplayText(deviceId, lines, durationSec = 30) {
    this.publish(
      `inkpad/${deviceId}/display/text`,
      { lines, duration_sec: durationSec },
      1,    // qos=1
      false,
    );
  }

  publishConfig(deviceId, config) {
    this.publish(
      `inkpad/${deviceId}/config/set`,
      config,
      1,    // qos=1
      false,
    );
  }
}

// Singleton
module.exports = new ClockMQTTClient();
